### PR TITLE
Add a versioning system

### PR DESCRIPTION
**Pull request by jayrbolton**
This is what I came up with regarding a system for url-based versioning of this api while keeping the code dry. New versions basically can deprecate or shadow the paths of old versions.

This has some breaking changes on the admin/config paths: eg. "/api/view/name" is now "/api/views?name=name". I'm positive that no code depends on these endpoints so I think it's safe. Except for changing paths from eg. "/api/query_results" to "/api/v1/query_results", all the rest of the api is the same as before.

And of course this will break clients that use the "api/path" calls currently (KE apps and landing page ui). We'll just have to update those. Going forward, this will give backwards compatibility for any modified endpoints.

### CODE REVIEW

**Comment by jayrbolton** [comment]
To add a new api version:
- make a new module under /src/relation_engine_server/api_modules called "api_v2.py"
- that module should export an "endpoints" and a "deprecations" dict
- import that module into server.py
- put the endpoints and deprecation dicts at the end of _API_VERSION and _API_DEPRECATION lists at the top of server.py
These versions are only incrementing single integers here, as i'm only concerned with breaking changes.

File: src/relation_engine_server/server.py [hunk 1]
```diff
@@ -25,137 +31,90 @@ def root():
...
+    version_int = _get_version(path_parts[0])
+    # Get the path and version number
+    api_path = '/'.join(path_parts[1:])
+    # Flag for whether the endpoint we find is deprecated
+    deprecated = False
+    # Find our method in the various versioned modules
+    # If it is not present in a later version, fall back to a previous version
+    # Iterates by starting at (version-1), stopping at 0, and stepping backwards
+    # Note: the mypy type checker has difficulties with the path_funcs dicts, so we ignore type checking below
+    for ver in range(version_int - 1, -1, -1):
```
Comment (JamesJeffryes) [inline src/relation_engine_server/server.py]: Hmm, I think this is to permit code reuse without boilerplate but I don't think I like the idea of automatically falling back to an older version of the api to find missing function. For example, we used to call the method "update_specs", "refresh_specs". With this versioning, I might forget that rename and accidentally call "refresh_specs" even though I intended to call "update_specs" with it's modern paramenters and behaviors, I just botched the name. I should get a  view_does_not_exist error not the older function if I make that call. Effectively this structure means that methods can be depreciated or redefined but never removed from future versions of an API

I also foresee an issue with maintenance. If old endpoints that are pulled though are explicitly specified in each version of the API, you can find the exact implementation for any API easily. with this, you need to look though the API files in reverse order until you find a method that matches just like this function.

**Comment by jayrbolton** [comment]
Good points, what if we made the api methods explicit, simply importing functions from previous versions, and eliminate that fallback system? That would simplify the logic in the route handler as well, so seems like a win-win, and will still keep the module code DRY

**Comment by JamesJeffryes** [comment]
Yup, it's a bit of boilerplate in each version of the API but I think it's worth it for simplicity and explicitness.

**Comment by jayrbolton** [comment]
:cool: made the above changes

**Comment by jayrbolton** [comment]
We couldjust add a deprecation header for all methods for any previous versions. We could even put a lifespan in there too and a have a global lifespan policy
We definitely need to do some serious coordination on the deploy for this. I'm good to just hold off for now.

**Pull request merged**